File: android/src/main/java/com/witsystem/top/flutterwitsystem/sdk/WitsSdkInit.java
package com.witsystem.top.flutterwitsystem.sdk;

import android.content.Context;

import com.witsystem.top.flutterwitsystem.add.AddDevice;
import com.witsystem.top.flutterwitsystem.add.ble.AddBleDevice;
import com.witsystem.top.flutterwitsystem.device.DeviceBasicInfo;
import com.witsystem.top.flutterwitsystem.device.DeviceManager;
import com.witsystem.top.flutterwitsystem.induce.Induce;
import com.witsystem.top.flutterwitsystem.induce.InduceUnlock;
import com.witsystem.top.flutterwitsystem.location.AppLocation;
import com.witsystem.top.flutterwitsystem.operation.BleOperation;
import com.witsystem.top.flutterwitsystem.operation.Operation;
import com.witsystem.top.flutterwitsystem.serialport.OpenSerialPort;
import com.witsystem.top.flutterwitsystem.serialport.SerialPort;
import com.witsystem.top.flutterwitsystem.smartconfig.SmartConfig;
import com.witsystem.top.flutterwitsystem.smartconfig.SmartConfigManager;
import com.witsystem.top.flutterwitsystem.unlock.BleUnlock;
import com.witsystem.top.flutterwitsystem.unlock.Unlock;

import java.util.List;


/**
 * SDK初始化
 */
public final class WitsSdkInit implements Register, WitsSdk {


    private Context context;

    private static Register mRegister;

    private String appId;

    private String userToken;

    private WitsSdkInit() {
    }

    public static Register getInstance() {
        if (mRegister == null) {
            synchronized (WitsSdkInit.class) {
                if (mRegister == null) {
                    mRegister = new WitsSdkInit();

                }
            }
        }
        return mRegister;
    }


    @Override
    public WitsSdk witsSdkInit(Context context, String appId, String userToken) {
        if (appId == null || userToken == null) {
            return null;
        }
        if (DeviceManager.getInstance(context, appId, userToken).getNetWorkDevice()) {//网络初始化结果
            this.context = context;
        }else if(DeviceManager.getInstance(context, appId, userToken).dataInitState()){//网络初始化失败在看本地初始化结果
            this.context = context;
        }
        this.appId = appId;
        this.userToken = userToken;
        if (context != null)
            AppLocation.startLocation(context);
        return this.context == null ? null : this;
    }


    @Override
    public List<DeviceBasicInfo> getDeviceInfo() {
        return DeviceManager.getInstance(context, null, null).getThreeDevices();
    }

    @Override
    public InduceUnlock getInduceUnlock() {
        return context == null ? null : Induce.instance(context);
    }

    @Override
    public BleUnlock getBleUnlock() {
        return Unlock.instance(context, appId, userToken);
    }

    @Override
    public Operation getOperation() {
        return BleOperation.instance(context, appId, userToken);
    }

    @Override
    public SerialPort getSerialPort() {
        return OpenSerialPort.instance(context);
    }

    @Override
    public AddDevice getAddBleDevice() {
        return AddBleDevice.instance(context, appId, userToken);
    }

    @Override
    public SmartConfig getSmartConfig() {
        return SmartConfigManager.getInstance(context,appId, userToken);
    }
}

File: ios/Classes/sdk/WitsSdkInit.swift
//
//
//  Created by yyjjls on 2020/4/30.
//

import Foundation

public class WitsSdkInit: NSObject, WitsSdk, Register {

    private var appId: String? = "";
    private var token: String? = "";
    private static var register: Register? = nil;
    private var ble: Ble;

     override init() {
        ble = Ble.getInstance;
        Induce.getInstance;
    }

    public static func getInstance() -> Register {
        if (register == nil) {
            register = WitsSdkInit();
        }
        return register!;
    }

    public func witsSdkInit(appId: String?, token: String?) -> WitsSdk? {
        self.appId = appId;
        self.token = token;
        let state = DeviceManager.getInstance(appId: appId!, token: token!).getNetWorkDevice();
        // 启动定位
        AppLocation.getInstance.startLocation();
        if (state) {
            return self
        } else if (DeviceManager.getInstance(appId: appId!, token: token!).dataInitState()) {
            return self;
        }
        return nil;
    }

    public func getDeviceInfo() -> [DeviceBasicInfo] {
        return DeviceManager.getInstance(appId: appId!, token: token!).getThreeDevices();
    }


    public func getInduceUnlock() -> Induce {
        return Induce.getInstance;
    }

    public func getBleUnlock() -> BleUnlock {
        return Unlock.getInstance(appId: appId, token: token);
    }


    public func getSerialPort() -> SerialPort {
        return OpenSerialPort.getInstance();
    }

    public func getAddBleDevice() -> AddDevice {
        return AddBleDevice.getInstance(appId: appId, token: token);
    }
}
